### ISSUE TITLE
Monitoring of registry.k8s.io

### ISSUE THREAD

**Issue by ameukam** [kubernetes/registry.k8s.io#17]
`archeio` is currently deployed in sandbox mode using Cloud Run: 

We have [built-in metrics](https://cloud.google.com/run/docs/monitoring) for Cloud Run but we may need some custom metrics to gain more visibility in the system we are building. 
Possible metrics to collect:
- HTTP code for every response from k8s.gcr.io

Alerting: https://github.com/kubernetes-sigs/oci-proxy/issues/75

**Comment by aojea** [comment]
Request:
- code and duration. Attributes version(v1 or v2 or unknown) and, if `path` is bounded or reasonable small (to avoid cardinality explosion) and know how many requests are used per docker image

Server:
- # active requests

**Comment by aojea** [comment]
maybe path, user-agent , ... and those things are better to log in an audit.log and not as metrics, so we can parse and report on them

Top images requested, ... by user agent,

**Comment by BobyMCbobs** [comment]
Added metrics in https://github.com/kubernetes-sigs/oci-proxy/pull/27

**Comment by BenTheElder** [comment]
Not sure how best to export it, but cloud run collects a lot of automatic trace data, see e.g.:
<img width="1637" alt="Screen Shot 2022-03-17 at 3 59 46 PM" src="https://user-images.githubusercontent.com/917931/158907803-5431f167-979c-450c-bcde-932703467c66.png">

I'm not sure we need custom metrics.


> maybe path, user-agent , ... and those things are better to log in an audit.log and not as metrics, so we can parse and report on them

These things are already logged AFAICT:
<img width="1584" alt="Screen Shot 2022-03-17 at 4 02 49 PM" src="https://user-images.githubusercontent.com/917931/158908058-982f8e08-0f66-48e9-9de0-e03fcd809e28.png">

I don't think they're going to be particularly useful though. The volume of requests is good to know, and at some point probably the source IPs, but for now we don't want to be in the business of logging / handling PII.

As far as user agent, we already need to be OCI compliant regardless of what clients are out there. We will be as compliant as GCR.

TLDR: IMO we need to track the request volumes we are dealing with, which is something we get OOTB.

We should look at how best to open this data to the community, carefully (because potential for PII, as before with k8s.gcr.io auditing).

**Comment by BenTheElder** [comment]
We have alerting in https://github.com/kubernetes/registry.k8s.io/issues/75#issuecomment-1236806204 (merging these)

We have some general metrics like QPS, memory and cpu usage, error rate, etc. out of the box by cloud run.

**Comment by BenTheElder** [comment]
Also noting: Kubernetes has a *ton* of CI doing things with container images all day long that will go red very quickly if registry.k8s.io is down and we'll catch on fairly quickly from that alone.

What we likely wouldn't notice as quickly from that though, is if we had a region specific issue. Most of our CI is us-central-1 on GCP. Some of it like kOps-AWS is actually running in ~random AWS regions globally and we might detect that there.

But we also have some cloud monitoring alerts for certificates and uptime since https://github.com/kubernetes/k8s.io/pull/4145

**Comment by BenTheElder** [comment]
> HTTP code for every response from k8s.gcr.io

Cloud Monitoring provides this and between GCP and AWS we have many dashboards tracking:
- HTTP request rates and response code / error rates
- Network traffic to/from the redirector
- Network traffic from S3
- Billing
- Overall request rates
- Region specific request rates
- Remaining egress rates from k8s.gcr.io
- CPU and Memory usage
- Request latencies

Plus detailed request logs, and automated uptime alerts to the SIG alerts mailinglist https://github.com/kubernetes/k8s.io/blob/main/infra/gcp/terraform/modules/oci-proxy/monitoring.tf
As well as alerts from e2e CI against the staging instance, and general Kubernetes CI alerts that will flag if the prod registry stops pulling since basically all of Kubernetes CI pulls from this.

So _monitoring_ is relatively complete.

https://github.com/kubernetes/registry.k8s.io/issues/89 is the issue to track analyzing metrics for analytics type purposes.
